examples/LabeledModels/LabeledModels.py

```diff
@@ -0,0 +1,564 @@
+
+# %%
+aNrm = DataArray(
+    mVec,
+    name="aNrm",
+    dims=("aNrm"),
+    attrs={"long_name": "Normalized Liquid Assets"},
+)
+post_state = Dataset({"aNrm": aNrm})
+
+print(post_state)
+
+# %% [markdown]
+# We can now define functions over the state space. In this basic model, we need action/policy/decision of consumption. Starting from the last period, we know that the solution is for the agent to consume all of its resources. Defining it as a function of the state space is easy.
+#
+
+# %%
+# optimal decision is to consume everything in the last period
+cNrm = DataArray(
+    mVec,
+    name="cNrm",
+    dims=state.dims,
+    coords=state.coords,
+    attrs={"long_name": "Consumption"},
+)
+actions = Dataset({"cNrm": cNrm})
+cNrm
+
+# %% [markdown]
+# ### The Value and Marginal Value functions.
+#
+
+# %% [markdown]
+# To define the value and marginal value functions in the last period, we can use the utility and marginal utility functions.
+#
+
+# %%
+v = u(cNrm)
+v.name = "v"
+v.attrs = {"long_name": "Value Function"}
+
+v_der = mu(cNrm)
+v_der.name = "v_der"
+v_der.attrs = {"long_name": "Marginal Value Function"}
+
+# %% [markdown]
+# It will also be useful to define the inverse value and inverse marginal value functions.
+#
+
+# %%
+v_inv = cNrm.copy()
+v_inv.name = "v_inv"
+v_inv.attrs = {"long_name": "Inverse Value Function"}
+
+v_der_inv = cNrm.copy()
+v_der_inv.name = "v_der_inv"
+v_der_inv.attrs = {"long_name": "Inverse Marginal Value Function"}
+
+# %% [markdown]
+# We can now create a xr.Dataset to store all of the variables/functions we have created. Datasets are useful containers of variables that are defined over the same dimensions, or in our case states.
+#
+
+# %%
+dataset = Dataset(
+    {
+        "cNrm": cNrm,
+        "v": v,
+        "v_der": v_der,
+        "v_inv": v_inv,
+        "v_der_inv": v_der_inv,
+    }
+)
+dataset
+
+# %% [markdown]
+# We can also create separate datasets for the value function variables and the policy function variable.
+#
+
+# %%
+value_function = Dataset(
+    {"v": v, "v_der": v_der, "v_inv": v_inv, "v_der_inv": v_der_inv}
+)
+policy_function = Dataset({"cNrm": cNrm})
+
+# %% [markdown]
+# Up to now I've used the word function for the variables stored as datasets. This is because using the `interp` method we can interpolate the values of the variables at any point in the state space. This is useful for solving the model, as we will see later.
+#
+
+# %%
+dataset.interp({"mNrm": np.sort(np.random.uniform(epsilon, 20, 10))})
+
+
+# %% [markdown]
+# Because of the curvature of the value and marginal value functions, it'll be useful to use the inverse value and marginal value functions instead and re-curve them. For this, I create a new class `ValueFunctionCRRA` that returns the appropriate value and marginal value functions.
+#
+
+# %%
+class ValueFunctionCRRA(object):
+    def __init__(self, dataset: xr.Dataset, CRRA: float):
+
+        self.dataset = dataset
+        self.CRRA = CRRA
```

Comment:
It seems like what you are doing here is making a callable wrapper around an xr.DataSet.

I wonder if this can be implemented in a more general way.

Reply:
There might be a more general way. This is to use the trick of using the inverse value function to approximate the value function.